src/Features/Core/Portable/EmbeddedLanguages/DateAndTime/DateAndTimeEmbeddedCompletionProvider.cs

```diff
@@ -0,0 +1,216 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Completion.Providers;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.DateAndTime;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Features.EmbeddedLanguages.DateAndTime
+{
+    internal partial class DateAndTimeEmbeddedCompletionProvider : CompletionProvider
+    {
+        private const string StartKey = nameof(StartKey);
+        private const string LengthKey = nameof(LengthKey);
+        private const string NewTextKey = nameof(NewTextKey);
+        private const string DescriptionKey = nameof(DescriptionKey);
+
+        // Always soft-select these completion items.  Also, never filter down.
+        private static readonly CompletionItemRules s_rules =
+            CompletionItemRules.Default.WithSelectionBehavior(CompletionItemSelectionBehavior.SoftSelection)
+                                       .WithFilterCharacterRule(CharacterSetModificationRule.Create(CharacterSetModificationKind.Replace, new char[] { }));
+
+        private readonly DateAndTimeEmbeddedLanguageFeatures _language;
+
+        public DateAndTimeEmbeddedCompletionProvider(DateAndTimeEmbeddedLanguageFeatures language)
+            => _language = language;
+
+        public override bool ShouldTriggerCompletion(SourceText text, int caretPosition, CompletionTrigger trigger, OptionSet options)
+        {
+            if (trigger.Kind == CompletionTriggerKind.Invoke ||
+                trigger.Kind == CompletionTriggerKind.InvokeAndCommitIfUnique)
+            {
+                return true;
+            }
+
+            if (trigger.Kind == CompletionTriggerKind.Insertion)
+            {
+                // We only trigger on typing if it's the first character in a sequence.
+                return char.IsLetter(trigger.Character) &&
+                       caretPosition >= 2 &&
+                       !char.IsLetter(text[caretPosition - 2]);
+            }
+
+            return false;
+        }
+
+        public override async Task ProvideCompletionsAsync(CompletionContext context)
+        {
+            if (!context.Options.GetOption(DateAndTimeOptions.ProvideDateAndTimeCompletions, context.Document.Project.Language))
```

Comment:
For some reason I can't reply to the original comment

> this is also an easy escape hatch for us. But i agree. we likely don't need a user visible UI option for this.

If we want an easy way to turn it off, should we just use a feature flag so we can turn it off remotely?